File: newton.cpp
// Compilar com:
// $ g++ -std=c++11 newton.cpp -o newton -O2 -larmadillo

/*
Initial:
   55.0000
  -12.5000
*/

#include <iostream>
#include <armadillo>
#include <math.h>

#define N 1000

int achou = 0;
arma::vec Initial;

arma::mat F(arma::vec X, int p){
    arma::mat Ret = {{{exp(-X[0]) + exp(X[1]) - pow(X[0],2) - 2*pow(X[1],3)}},
                     {{cos(X[0] + p*X[1]) + p*X[0] - X[1] - 1}}};
    return Ret;
}

arma::mat JF(arma::vec X, int p){
    arma::mat Ret = {{{-2*X[0] - exp(-X[0])},{exp(X[1]) - 6*pow(X[1],2)}},
                     {{p - sin(X[0] + p*X[1])},{-p * sin(X[0] + p*X[1]) - 1}}};
    return Ret;
}

arma::mat newton(arma::vec X,double tol, int p){

    int k = 0;

    while (k < N){

        k++;

        arma::vec Delta = (-arma::inv(JF(X,p))) * F(X,p);
        if(Delta.has_nan()) break;

        Delta = arma::solve(JF(X,p),-F(X,p));
        X += Delta;

        if (arma::norm(Delta,"inf") < tol){
            if(Delta.has_nan() == false){
                //achou = 1;
                Initial.print("Initial:");
                Delta.print("Delta:");
                X.print("X:");
                F(X,21).print("F(X) com p=21:");
                std::cout << "K = " << k << std::endl;
            }
            return X;
        }
    }

    return {0,0};
}

int main(){

    arma::mat Sol;

    Initial = {0.16,1.31};
    Sol = newton(Initial,0.00001,9);
    Initial.print("Initial:");
    Sol.print("Solucao com p=9:");
    F(Sol,9).print("F(Sol) com p=9:");

    Initial = {0.7, 6.14};
    Sol = newton(Initial,0.00001,9);
    Initial.print("Initial:");
    Sol.print("Solucao com p=9:");
    F(Sol,9).print("F(Sol) com p=9:");

    for(double i=1000;i<100000;i+=10)
        for(double j=1000;j<100000;j+=10){
            Initial = {1.0*i,1.0*j};
            Sol = newton(Initial,0.00001,21);
        }

    Initial.print("Initial:");
    Sol.print("Solucao com p=21:");
    F(Sol,21).print("F(Sol) com p=21:");

    return 0;
}

File: newton.py
import numpy

# Altere o p aqui conforme desejado. Gabriela: p = 9, Luiz: p = 21
p = 9
chute = numpy.zeros(2)

def chutesQueDivergem():
    i = 0
    while (i < 25):
        j = 0
        while (j < 25):
            chute[0] = i
            chute[1] = j
            if (newton(chute,0.00001,1000) is None):
                print "Diverge: " + str(chute)
            j = j + 0.25
        i = i + 0.25

def chutesParaSol1():

    sol1 = numpy.zeros(2)
    if(p == 9):
        sol1[0] = 0.16
        sol1[1] = 1.31
    elif(p == 21):
        sol1[0] = 0.15
        sol1[1] = 1.32

    i = 0
    while (i < 25):
        j = 0
        while (j < 25):
            chute[0] = i
            chute[1] = j
            resultado = newton(chute,0.00001,5,nFixo=True)
            if(resultado is not None and (numpy.linalg.norm(resultado - sol1,numpy.inf) < 0.01)):
                print "Converge para a solucao " + str(sol1) + ": " + str(chute)
            j = j + 0.25
        i = i + 0.25

def chutesParaSol2():

    sol2 = numpy.zeros(2)
    if(p == 9):
        sol2[0] = 0.7
        sol2[1] = 6.14
    elif(p == 21):
        sol2[0] = 0.38
        sol2[1] = 6.13

    i = 0
    while (i < 25):
        j = 0
        while (j < 25):
            chute[0] = i
            chute[1] = j
            resultado = newton(chute,0.00001,5,nFixo=True)
            if(resultado is not None and (numpy.linalg.norm(resultado - sol2,numpy.inf) < 0.01)):
                print "Converge para a solucao " + str(sol2) + ": " + str(chute)
            j = j + 0.25
        i = i + 0.25

def F(X):
    ret = numpy.zeros(2)

    ret[0] = numpy.exp(-X[0]) + numpy.exp(X[1]) - pow(X[0],2) - 2*pow(X[1],3)
    ret[1] = numpy.cos(X[0] + p*X[1]) + p*X[0] - X[1] - 1

    return ret


def JF(X):
    ret = numpy.zeros((2,2))

    ret[0,0] = -2*X[0] - numpy.exp(-X[0])
    ret[0,1] = numpy.exp(X[1]) - 6*pow(X[1],2)
    ret[1,0] = p - numpy.sin(X[0] + p*X[1])
    ret[1,1] = -p * numpy.sin(X[0] + p*X[1]) - 1

    return ret


def newton(X,tol,n,verboso=False,nFixo=False):

    k = 0;

    while (k < n):
        k = k + 1
        delta = numpy.dot((-numpy.linalg.inv(JF(X))),F(X))

        if(numpy.isnan(numpy.sum(delta))):
            break

        X = X + delta

        if ((numpy.linalg.norm(delta,numpy.inf) < tol)):
            if(verboso):
                print "chute:"
                print chute
                print "Delta"
                print delta
                print "X:"
                print X
                print "F(X) com p=" + str(p) + ":"
                print F(X)

            if(k==n):
                return X
            if not (nFixo):
                return X

    return None

chutesParaSol1()
print "--------------------------------------------"
chutesParaSol2()
print "--------------------------------------------"
chutesQueDivergem()
